# 第2章: UNIXコマンドの基礎

## 10. 行数のカウント
行数をカウントせよ．確認にはwcコマンドを用いよ．

In [2]:
cat hightemp.txt | wc -l

      24


### python実装

In [5]:
with open('hightemp.txt', 'r') as f:
    print(len([line for line in f]))

24


## 11. タブをスペースに置換
タブ1文字につきスペース1文字に置換せよ．確認にはsedコマンド，trコマンド，もしくはexpandコマンドを用いよ．

In [3]:
cat hightemp.txt | tr '\t' ' '

高知県 江川崎 41 2013-08-12
埼玉県 熊谷 40.9 2007-08-16
岐阜県 多治見 40.9 2007-08-16
山形県 山形 40.8 1933-07-25
山梨県 甲府 40.7 2013-08-10
和歌山県 かつらぎ 40.6 1994-08-08
静岡県 天竜 40.6 1994-08-04
山梨県 勝沼 40.5 2013-08-10
埼玉県 越谷 40.4 2007-08-16
群馬県 館林 40.3 2007-08-16
群馬県 上里見 40.3 1998-07-04
愛知県 愛西 40.3 1994-08-05
千葉県 牛久 40.2 2004-07-20
静岡県 佐久間 40.2 2001-07-24
愛媛県 宇和島 40.2 1927-07-22
山形県 酒田 40.1 1978-08-03
岐阜県 美濃 40 2007-08-16
群馬県 前橋 40 2001-07-24
千葉県 茂原 39.9 2013-08-11
埼玉県 鳩山 39.9 1997-07-05
大阪府 豊中 39.9 1994-08-08
山梨県 大月 39.9 1990-07-19
山形県 鶴岡 39.9 1978-08-03
愛知県 名古屋 39.9 1942-08-02


### python実装

In [7]:
with open('hightemp.txt', 'r') as f:
    for line in f:
        print(line.rstrip().replace("\t", " "))

高知県 江川崎 41 2013-08-12
埼玉県 熊谷 40.9 2007-08-16
岐阜県 多治見 40.9 2007-08-16
山形県 山形 40.8 1933-07-25
山梨県 甲府 40.7 2013-08-10
和歌山県 かつらぎ 40.6 1994-08-08
静岡県 天竜 40.6 1994-08-04
山梨県 勝沼 40.5 2013-08-10
埼玉県 越谷 40.4 2007-08-16
群馬県 館林 40.3 2007-08-16
群馬県 上里見 40.3 1998-07-04
愛知県 愛西 40.3 1994-08-05
千葉県 牛久 40.2 2004-07-20
静岡県 佐久間 40.2 2001-07-24
愛媛県 宇和島 40.2 1927-07-22
山形県 酒田 40.1 1978-08-03
岐阜県 美濃 40 2007-08-16
群馬県 前橋 40 2001-07-24
千葉県 茂原 39.9 2013-08-11
埼玉県 鳩山 39.9 1997-07-05
大阪府 豊中 39.9 1994-08-08
山梨県 大月 39.9 1990-07-19
山形県 鶴岡 39.9 1978-08-03
愛知県 名古屋 39.9 1942-08-02


## 12. 1列目をcol1.txtに，2列目をcol2.txtに保存
各行の1列目だけを抜き出したものをcol1.txtに，2列目だけを抜き出したものをcol2.txtとしてファイルに保存せよ．確認にはcutコマンドを用いよ．

**TODO**unix cutコマンドで２つのファイルを同時に書き込む方法が思い浮かばなかった　標準出力と標準エラー出力にそれぞれ吐き分けられる？

In [10]:
with open("hightemp.txt", "r") as read_f:
    with open("col1.txt", "w") as write_f_1:
        with open("col2.txt", "w") as write_f_2:
            for line in read_f:
                line_elems = line.rstrip().split("\t")
                write_f_1.write(line_elems[0] + "\n")
                write_f_2.write(line_elems[1] + "\n")

In [11]:
cat col1.txt

高知県
埼玉県
岐阜県
山形県
山梨県
和歌山県
静岡県
山梨県
埼玉県
群馬県
群馬県
愛知県
千葉県
静岡県
愛媛県
山形県
岐阜県
群馬県
千葉県
埼玉県
大阪府
山梨県
山形県
愛知県


In [12]:
cat col2.txt

江川崎
熊谷
多治見
山形
甲府
かつらぎ
天竜
勝沼
越谷
館林
上里見
愛西
牛久
佐久間
宇和島
酒田
美濃
前橋
茂原
鳩山
豊中
大月
鶴岡
名古屋


## 13. col1.txtとcol2.txtをマージ
12で作ったcol1.txtとcol2.txtを結合し，元のファイルの1列目と2列目をタブ区切りで並べたテキストファイルを作成せよ．確認にはpasteコマンドを用いよ．

paste -d "\t" col1.txt col2.txt 

### python実装

In [15]:
with open("col1.txt", "r") as read_f_1, open("col2.txt", "r") as read_f_2, open("concate.txt", "w") as write_f:
    for line_1, line_2 in zip(read_f_1, read_f_2):
        write_f.write(line_1.rstrip() + "\t" + line_2)

In [16]:
cat concate.txt

高知県	江川崎
埼玉県	熊谷
岐阜県	多治見
山形県	山形
山梨県	甲府
和歌山県	かつらぎ
静岡県	天竜
山梨県	勝沼
埼玉県	越谷
群馬県	館林
群馬県	上里見
愛知県	愛西
千葉県	牛久
静岡県	佐久間
愛媛県	宇和島
山形県	酒田
岐阜県	美濃
群馬県	前橋
千葉県	茂原
埼玉県	鳩山
大阪府	豊中
山梨県	大月
山形県	鶴岡
愛知県	名古屋


## 14. 先頭からN行を出力
自然数Nをコマンドライン引数などの手段で受け取り，入力のうち先頭のN行だけを表示せよ．確認にはheadコマンドを用いよ．

### headコマンドの実行結果

$ head -n5 hightemp.txt 

高知県	江川崎	41	2013-08-12  
埼玉県	熊谷	40.9	2007-08-16  
岐阜県	多治見	40.9	2007-08-16  
山形県	山形	40.8	1933-07-25  
山梨県	甲府	40.7	2013-08-10  


下記コードをhead.pyとして保存

In [23]:
#!/Users/kzfm/.pyenv/shims/python 
import sys

args = sys.argv

if len(args)!=3:
    print("""This program is required 2 argments!""")
else:
    with open(args[2], 'r') as f:
        count = int(args[1])
        for line in f:
            print(line, end="")
            count -= 1
            if count == 0:
                break

ValueError: invalid literal for int() with base 10: '-f'

### 実行結果
chapter2 $python head.py 5 hightemp.txt   
高知県	江川崎	41	2013-08-12  
埼玉県	熊谷	40.9	2007-08-16  
岐阜県	多治見	40.9	2007-08-16  
山形県	山形	40.8	1933-07-25  
山梨県	甲府	40.7	2013-08-10  

## 15. 末尾のN行を出力
自然数Nをコマンドライン引数などの手段で受け取り，入力のうち末尾のN行だけを表示せよ．確認にはtailコマンドを用いよ．

### tailコマンドの実行結果¶
$tail -n 5 hightemp.txt 
埼玉県	鳩山	39.9	1997-07-05
大阪府	豊中	39.9	1994-08-08
山梨県	大月	39.9	1990-07-19
山形県	鶴岡	39.9	1978-08-03
愛知県	名古屋	39.9	1942-08-02

下記コードをtail.pyとして保存

In [24]:
#!/Users/kzfm/.pyenv/shims/python 
import sys

args = sys.argv

if len(args)!=3:
    print("""This program is required 2 argments!""")
else:
    with open(args[2], 'r') as f:
        line_list = f.readlines()
        len_list = len(line_list)
        N = int(args[1])
        for i in range(N):
            print(line_list[len_list-N+i], end="")

ValueError: invalid literal for int() with base 10: '-f'

### 実行結果

chapter2 $python tail.py 5 hightemp.txt  
埼玉県	鳩山	39.9	1997-07-05  
大阪府	豊中	39.9	1994-08-08  
山梨県	大月	39.9	1990-07-19  
山形県	鶴岡	39.9	1978-08-03  
愛知県	名古屋	39.9	1942-08-02  

## 16. ファイルをN分割する
自然数Nをコマンドライン引数などの手段で受け取り，入力のファイルを行単位でN分割せよ．同様の処理をsplitコマンドで実現せよ．

### splitコマンド
split -l 5 hightemp.txt

### python実装

In [25]:
#!/Users/kzfm/.pyenv/shims/python 
import sys

args = sys.argv

if len(args)!=3:
    print("""This program is required 2 argments!""")
else:
    with open(args[2], 'r') as f:
        line_list = f.readlines()
        len_list = len(line_list)
        N = int(args[1])
        for i in range(len_list//N):
            with open(str(i) + '.txt', 'w') as f:
                f.write("".join(line_list[i*N:(i+1)*N]))
        else:
            num = i+1
        with open(str(num) + '.txt', 'w') as f:
            f.write("".join(line_list[N * num:]))

ValueError: invalid literal for int() with base 10: '-f'

## 17. １列目の文字列の異なり
1列目の文字列の種類（異なる文字列の集合）を求めよ．確認にはsort, uniqコマンドを用いよ．

cut -c1  hightemp.txt | sort |   
千  
和  
埼  
大  
山  
岐  
愛  
群  
静  
高  

In [31]:
def get_unique_initial_char(filepath):
    with open(filepath, 'r') as f:
        char_set = set()
        for line in f:
            char_set.add(line[0])
    return char_set

In [32]:
get_unique_initial_char('hightemp.txt')

{'千', '和', '埼', '大', '山', '岐', '愛', '群', '静', '高'}

## 18. 各行を3コラム目の数値の降順にソート
各行を3コラム目の数値の逆順で整列せよ（注意: 各行の内容は変更せずに並び替えよ）．確認にはsortコマンドを用いよ（この問題はコマンドで実行した時の結果と合わなくてもよい）．

### unixコマンド
sort -k 3 hightemp.txt | tail -r

In [37]:
temp_dir = {10: "a", 30 : "b", 20 : "c", 15 : "d"}

for kv in sorted(temp_dir.items(), key=lambda x: x[0]):
    print(kv[0], kv[1])

10 a
15 d
20 c
30 b


In [44]:
def get_unique_initial_char(filepath):
    # ３カラム目と行の文字列を取り込む
    with open(filepath, 'r') as f:
        temp_dir = {}
        for line in f:
            temprature = line.split("\t")[2]
            temp_dir[line.rstrip()] = float(temprature)
    
    for kv in reversed(sorted(temp_dir.items(), key=lambda x: x[1])):
        print(kv[0])

In [45]:
get_unique_initial_char("hightemp.txt")

高知県	江川崎	41	2013-08-12
岐阜県	多治見	40.9	2007-08-16
埼玉県	熊谷	40.9	2007-08-16
山形県	山形	40.8	1933-07-25
山梨県	甲府	40.7	2013-08-10
静岡県	天竜	40.6	1994-08-04
和歌山県	かつらぎ	40.6	1994-08-08
山梨県	勝沼	40.5	2013-08-10
埼玉県	越谷	40.4	2007-08-16
愛知県	愛西	40.3	1994-08-05
群馬県	上里見	40.3	1998-07-04
群馬県	館林	40.3	2007-08-16
愛媛県	宇和島	40.2	1927-07-22
静岡県	佐久間	40.2	2001-07-24
千葉県	牛久	40.2	2004-07-20
山形県	酒田	40.1	1978-08-03
群馬県	前橋	40	2001-07-24
岐阜県	美濃	40	2007-08-16
愛知県	名古屋	39.9	1942-08-02
山形県	鶴岡	39.9	1978-08-03
山梨県	大月	39.9	1990-07-19
大阪府	豊中	39.9	1994-08-08
埼玉県	鳩山	39.9	1997-07-05
千葉県	茂原	39.9	2013-08-11


## 19. 各行の1コラム目の文字列の出現頻度を求め，出現頻度の高い順に並べる
各行の1列目の文字列の出現頻度を求め，その高い順に並べて表示せよ．確認にはcut, uniq, sortコマンドを用いよ．

In [48]:
def print_ken_freq(filepath):
    with open(filepath, 'r') as f:
        ken_dir = {}
        for line in f:
            ken = line.split("\t")[0]
            if ken in ken_dir:
                ken_dir[ken] += 1
            else:
                ken_dir[ken] = 1
    
    for kv in reversed(sorted(ken_dir.items(), key=lambda x: x[1])):
        print(kv[0])

In [49]:
print_ken_freq("hightemp.txt")

群馬県
山梨県
山形県
埼玉県
千葉県
愛知県
静岡県
岐阜県
大阪府
愛媛県
和歌山県
高知県


### unixコマンドでの確認

cut -f1 hightemp.txt | sort | uniq -c| sort | tail -r  
   3 群馬県  
   3 山梨県  
   3 山形県  
   3 埼玉県  
   2 静岡県  
   2 愛知県  
   2 岐阜県  
   2 千葉県  
   1 和歌山県  
   1 高知県  
   1 愛媛県  
   1 大阪府  
  